src/main/java/com/glodblock/github/extendedae/common/me/taglist/TagPriorityList.java

```diff
@@ -45,27 +68,174 @@ public Iterable<AEKey> getItems() {
 
     @SuppressWarnings("deprecation")
     private boolean eval(@NotNull Object obj) {
+        if (DEBUG_ENABLED) {
+            LOGGER.info("Evaluating object: {}", obj);
+        }
+        
+        // First check if we have an empty taglist or a taglist with only operators that should match nothing
+        if (parsedExpression instanceof TagExpParser.EmptyExpression) {
+            if (DEBUG_ENABLED) {
+                LOGGER.info("Empty expression, returning false immediately");
+            }
+            return false;
+        }
+        
         Holder<?> refer = null;
         if (obj instanceof Item item) {
             refer = item.builtInRegistryHolder();
+            if (DEBUG_ENABLED) {
+                LOGGER.info("Object is an Item: {}", item.getDescriptionId());
+            }
         } else if (obj instanceof Fluid fluid) {
             refer = fluid.builtInRegistryHolder();
+            if (DEBUG_ENABLED) {
+                LOGGER.info("Object is a Fluid: {}", fluid.toString());
+            }
         }
-        if (refer != null) {
-            boolean pass = true;
-
-            if (!whiteSet.isEmpty()) {
-                pass = refer.tags().anyMatch(whiteSet::contains);
+        
+        if (refer == null) {
+            if (DEBUG_ENABLED) {
+                LOGGER.info("Object is neither Item nor Fluid, returning false");
             }
-
-            if (pass) {
-                if (!blackSet.isEmpty()) {
-                    return refer.tags().noneMatch(blackSet::contains);
+            return false;
+        }
+        
+        // Collect the object's tags for logging
+        Set<TagKey<?>> objectTags = new HashSet<>();
+        refer.tags().forEach(objectTags::add);
+        
+        if (DEBUG_ENABLED) {
+            LOGGER.info("Object has {} tags: {}", objectTags.size(), formatTags(objectTags));
+        }
+        
+        // Special case: Detect if we're dealing with an AND expression with direct tag expressions
+        if (parsedExpression instanceof TagExpParser.AndExpression andExpression) {
```

Comment:
What will happen if it is a nested AND expression? like `(exp1 & exp2) | (exp3 & exp4)`

Reply:
@GlodBlock it works now :) Just pushed change to update that and have more robust support.

Input chest:
![image](https://github.com/user-attachments/assets/918eafdc-b867-4f2c-bbf4-f76594d19491)
![image](https://github.com/user-attachments/assets/3ac3bcbf-d3cf-4dee-9230-b742db6d6d3c)

Filter (`(c:tools) & !(c:tools/paxel)`):
![image](https://github.com/user-attachments/assets/2b019311-c063-4a8e-b6ba-1644554eaf0c)


Output:
![image](https://github.com/user-attachments/assets/5dd0f12b-7737-447e-bb76-3b3e23a22c1b)